hack/mock-unittest-data/yarn.py

```diff
@@ -0,0 +1,132 @@
+
+    def has_path_but_no_checksum(pkg: dict[str, Any]) -> bool:
+        cache = pkg["children"]["Cache"]
+        return cache["Checksum"] is None and cache["Path"] is not None
+
+    pkgchain = itertools.chain(
+        # take 1 or 2 examples of each supported protocol
+        _filter_pkgs_by_pattern("@npm:", yarn_packages_no_git_protocol, 1),
+        _filter_pkgs_by_pattern("@workspace:", yarn_packages_no_git_protocol, 2),
+        _filter_pkgs_by_pattern("@patch:", yarn_packages_no_git_protocol, 2),
+        _filter_pkgs_by_pattern("@file:", yarn_packages_no_git_protocol, 2),
+        _filter_pkgs_by_pattern("@portal:", yarn_packages_no_git_protocol, 1),
+        _filter_pkgs_by_pattern("@link:", yarn_packages_no_git_protocol, 1),
+        _filter_pkgs_by_pattern("@https:.*tar.gz", yarn_packages_no_git_protocol, 1),
+        # make sure to include the curious case where checksum is null but path isn't
+        _filter_pkgs(has_path_but_no_checksum, yarn_packages_no_git_protocol),
+    )
+
+    # sort packages and make them unique by "value" (the locator)
+    pkgdict = {pkg["value"]: pkg for pkg in pkgchain}
+    selected_packages = [pkg for _, pkg in sorted(pkgdict.items())]
+
+    for pkg in selected_packages:
+        # drop large unused Dependencies attribute
+        pkg["children"].pop("Dependencies", None)
+        # replace the repo directory path in Cache.Path with a placeholder
+        if cache_path := pkg["children"]["Cache"]["Path"]:
+            path_with_placeholder = Path("{repo_dir}", Path(cache_path).relative_to(repo_dir))
+            pkg["children"]["Cache"]["Path"] = path_with_placeholder.as_posix()
+
+    return selected_packages
+
+
+def _filter_pkgs(
+    predicate: Callable[[dict[str, Any]], bool],
+    pkgs: list[dict[str, Any]],
+    max_items: Optional[int] = None,
+) -> Iterator[dict[str, Any]]:
+    filtered = filter(predicate, pkgs)
+    if max_items:
+        return itertools.islice(filtered, max_items)
+    else:
+        return filtered
+
+
+def _filter_pkgs_by_pattern(
+    pattern: str,
+    pkgs: list[dict[str, Any]],
+    max_items: Optional[int] = None,
+) -> Iterator[dict[str, Any]]:
+    # check if the locator contains the regex pattern
+    def matches_pattern(pkg: dict[str, Any]) -> bool:
+        return re.search(pattern, pkg["value"]) is not None
+
+    return _filter_pkgs(matches_pattern, pkgs, max_items)
+
+
+def main() -> None:
+    print_banner("Generating mock data for yarn unit tests")
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        repo_dir = clone_repo(Path(tmpdir))
+        yarninfo_output = run_yarninfo(repo_dir)
+
+    selected_packages = process_yarninfo(yarninfo_output, repo_dir)
+
+    print_banner(
+        "You can copy the following to tests/unit/package_managers/yarn/test_resolver.py "
+        "(you will need to re-format it with 'black')"
+    )
+
+    pprint.pprint(selected_packages, sort_dicts=False)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Good point, we do.

Reply:
For my information: why?

Reply:
Technically, you may have omitted it here because you let the script to end with a backtrace on error rather than handle the exceptions and produce user-friendly errors. It's just a common practice to add that specific construction for standalone scripts which do produce user-friendly errors, because in such cases you exit the script with a non-zero error code which would otherwise be 0 all the time since that's what the python runtime would interpret None as. At least that's my take on it, @ben-alkov might have something else, but for me it's ^this.

Reply:
Right, if we had a `main() -> int` that would make sense, but this one is `main() -> None`

Reply:
but I can add that in case it changes in the future

Reply:
IIUC (maybe mot), in practice `main() -> None` == 0 (to the controlling process e.g. bash)